src/dataflow/src/render/upsert.rs

```diff
@@ -482,20 +510,19 @@ where
                                     };
 
                                     if let Some(old_value) = old_value {
-                                        // retract old value
-                                        session.give((old_value, cap.time().clone(), -1));
+                                        // Ensure we put the source in a permanently error'd state
```

Comment:
I was a bit hard to convince myself that the change does what it says because we just remove the `Some(Err(err))` branch and the "placeholder" value up above is not really a placeholder value, right? That part up above pipes the key decoding error into the `decoded_value`, so that one `session.give()` below will emit the key decoding error that happens to be in `new_value`.

We should change the comment up above about it being a placeholder, and embellish this comment here to say that we do emit errors for keys, but then never retract them.

Reply:
Not saying the code is wrong, it was just hard for me to follow the flow of the errors! :sweat_smile: